src/hotspot/share/prims/vectorSupport.cpp

```diff
@@ -459,19 +459,50 @@ int VectorSupport::vop2ideal(jint id, BasicType bt) {
 }
 #endif // COMPILER2
 
+int VectorSupport::get_max_multifield_count(const Symbol* payload_name) {
+  assert(payload_name, "");
+  vmSymbolID sid = vmSymbols::find_sid(payload_name);
+  int size = VM_Version::max_vector_size(T_BYTE);
+  switch(sid) {
+    case vmSymbolID::jdk_internal_vm_vector_VectorPayloadMFMaxB_enum:
+    case vmSymbolID::jdk_internal_vm_vector_VectorPayloadMFMaxBZ_enum:
+    case vmSymbolID::jdk_internal_vm_vector_VectorPayloadMFMaxBB_enum:
+       return size;
+    case vmSymbolID::jdk_internal_vm_vector_VectorPayloadMFMaxS_enum:
+    case vmSymbolID::jdk_internal_vm_vector_VectorPayloadMFMaxSZ_enum:
+    case vmSymbolID::jdk_internal_vm_vector_VectorPayloadMFMaxSB_enum:
+       return size / type2aelembytes(T_SHORT);
+    case vmSymbolID::jdk_internal_vm_vector_VectorPayloadMFMaxI_enum:
+    case vmSymbolID::jdk_internal_vm_vector_VectorPayloadMFMaxIZ_enum:
+    case vmSymbolID::jdk_internal_vm_vector_VectorPayloadMFMaxIB_enum:
+       return size / type2aelembytes(T_INT);
+    case vmSymbolID::jdk_internal_vm_vector_VectorPayloadMFMaxL_enum:
+    case vmSymbolID::jdk_internal_vm_vector_VectorPayloadMFMaxLZ_enum:
+    case vmSymbolID::jdk_internal_vm_vector_VectorPayloadMFMaxLB_enum:
+       return size / type2aelembytes(T_LONG);
+    case vmSymbolID::jdk_internal_vm_vector_VectorPayloadMFMaxF_enum:
+       return size / type2aelembytes(T_FLOAT);
+    case vmSymbolID::jdk_internal_vm_vector_VectorPayloadMFMaxD_enum:
+       return size / type2aelembytes(T_DOUBLE);
+    default  : fatal("unknown symbol"); return -1;
+  }
+}
+
+int VectorSupport::max_vector_size(BasicType bt) {
+  if (is_java_primitive(bt)) {
+    return VM_Version::max_vector_size(bt);
```

Comment:
`VM_Version::max_vector_size(bt)` is a platform specific method, and currently only x86 contains it. So this makes the jdk image build failure on other platforms like aarch64. Could we define a public/common method and implement it on each platform like matcher?

Reply:
Correct, we need handle it for aarch64.  Matcher routines will not be accessible by C1 only builds. That was a limitation in existing [_VectorSupport_GetMaxLaneCount_](https://github.com/openjdk/valhalla/blob/lworld/src/hotspot/share/prims/vectorSupport.cpp#L583) functionality. Max vector size is tightly bound to targets.

> Could we define a public/common method and implement it on each platform like matcher

Yes, x86 already does that currently, we need to have a target specific handling one way or other, We need need max vector size to estimate bundle size during field parsing and wanted to avoid its dependency on opto compiler.

Reply:
Yeah, I agree that we need a non-c2 specific method to get the max vector size. But it is added for x86 now. And I can help add the aarch64 part if you need. My concern is for other platforms that do not have such a method, the jdk image will build failure.